compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLinkExecutables.cpp

```diff
@@ -40,22 +60,110 @@ struct SPIRVLinkExecutablesPass final
     }
     sourceExecutableOps.resize(retainSize);
 
-    // Guess a module name, if needed, to make the output files readable.
-    std::string moduleName = guessModuleName(moduleOp, "spirv_module");
+    // Note that at runtime, for a particular executable, only one variant of it
+    // will be loaded. So, all variants of an executable are expected to provide
+    // the exact same set of entry points; this way we can guarantee no matter
+    // which variant is chosen, we have all entry points to call into. The same
+    // entry point in different variants may have different target requirements
+    // though.
+    //
+    // The input to the linking stage are a collection of executables, each may
+    // have multiple variants, but only ever provide one entry point. Together
+    // with the above restriction, we can link two executables if and only if
+    // their variants have the exact same set of target requirements. Under such
```

Comment:
Are we expecting to do VCE simplification at the variant level after linking? Or is that too optimistic to try to allow a model to run on unspecified hardware out of pure chance.

Reply:
Oh I think I see that's what https://github.com/openxla/iree/pull/15888#issuecomment-1853102882 is

Reply:
Yeah that's the plan. We replace the provided target environment with required ones, and then turn them into device queries and unique based on that. See https://github.com/openxla/iree/pull/15977/files#diff-c1bc675af2875eab38eb993022440b69b51e2f9b6a3197871980f2a45810040e.